### PR TITLE
fix: enforce ExactPrice plugin in on_buy_plugin and cap overpayment in buy()

### PR DESCRIPTION
**Pull request by 2xburnt**
## Summary

Defense-in-depth hardening for the asset contract buy flow.

### 1. Load ExactPrice plugin in `on_buy_plugin` (`plugin.rs`)
The existing comment states ExactPrice should run before other buy plugins, but it was not being loaded. Now it is loaded and executed first.

### 2. Cap payment to listing price in `buy()` (`sale.rs`)
Regardless of plugin configuration, `buy()` now caps the payment to the listing price and refunds any excess to the buyer via `BankMsg::Send`. The seller never silently receives more than the listing price.

## Testing
All 44 existing tests pass. No behavioral change for correctly-priced purchases.

## Summary by CodeRabbit

* **New Features**
  * Added automatic refunds when buyers pay more than the listed price.
  * Added exact-price validation for purchases.
  * Royalty calculations now use the listing price and require a valid, nonzero ask price.
  * Purchase processing now validates the configured exact-price rules alongside other payment checks.

* **Bug Fixes**
  * Prevented excess payments from being incorrectly included in seller or royalty calculations.
  * Improved handling of missing or mismatched payment funds.

<h3>Greptile Summary</h3>

This change ensures asset purchases settle at the listing price: excess payment is refunded to the buyer, royalty deductions are calculated from the listed price, and ExactPrice rejects amounts that differ from the listing price.

The concern that refund, seller settlement, and royalty transfers could leave the contract unable to fund a later transfer was disproved by an executed royalty-enabled purchase. A buyer sent 120 uxion for a 100 uxion listing with a 50 uxion royalty. The contract transferred 20 uxion back to the buyer, 50 uxion to the seller, and 50 uxion to the artist, ending with a zero balance. The complete Rust workspace test suite also passed.

<h3>Confidence Score: 5/5</h3>

The reviewed payment and royalty path is safe to merge based on the exercised purchase flow and passing workspace tests.

The final review contains no unresolved defects. The executed contract flow directly verified the transfer ordering and balances for an overpayment combined with a royalty deduction.

**Files Needing Attention:** No files require follow-up from this review. The validated behavior spans contracts/asset/src/execute/sale.rs, contracts/asset/src/plugin.rs, and contracts/asset/src/default_plugins.rs.

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- I examined the contract code paths for purchase, refunds, royalties, and seller settlements across asset/plugin and sale execution.
- I ran focused cargo tests that exercise the overpayment royalty flow and confirmed the funds are fully provided in sequence: 120 units before responses, 100 after refund, 50 after seller settlement, and 0 after royalty payment; the observed balance changes matched the expected PRE and POST states.
- I ran the workspace test suite and it completed successfully.

<a href="https://app.greptile.com/trex/runs/20029805/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<sub>Reviews (1): Last reviewed commit: ["style: inline format args in plugin test"](https://github.com/burnt-labs/contracts/commit/ac33bf348478a1220f055edb7d508e52ba0f1ac2) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=55576229)</sub>

### CODE REVIEW

**Review by crucible-burnt — COMMENTED** [review]
## 🔍 Crucible Security Review

### Summary
This PR adds defense-in-depth overpayment protection in `buy()` and ensures the ExactPrice plugin is enforced in `on_buy_plugin`. These changes relate to marketplace payment handling security.

### Security Assessment
- **Risk Level**: **Low** (improvements with no regressions)

**✅ Positive Findings:**

1. **Overpayment Refund Defense** (`sale.rs:60-75`)
   - Caps payment to listing price regardless of ExactPrice plugin configuration
   - Refunds excess to buyer automatically
   - Uses `checked_sub` to prevent underflow
   - Good comment explaining the defense-in-depth rationale

2. **ExactPrice Plugin Enforcement** (`plugin.rs:382-414`)
   - Now loads and runs ExactPrice plugin in `on_buy_plugin`
   - Runs BEFORE AllowedCurrencies plugin
   - Ensures consistent enforcement across all buy paths

**⚠️ Observations:**

1. **Refund Message Ordering**
   - Refund is added via `response.add_message()` - confirm execution order is correct (refund should happen after successful purchase)
   - Consider: should refund fail the tx or proceed with capped payment?

2. **Error Type**
   - `ContractError::InsufficientFunds` on line 63 is used for checked_sub error - semantically this is a math error, not insufficient funds. Minor but could confuse debugging.

### Immunefi Pattern Check
- **#65150 (Marketplace PendingSale expiry)**: ⚠️ **Not addressed by this PR** - this PR fixes overpayment handling but the core issue of expired PendingSale approvals allowing NFT theft appears to be addressed by PR #118 ("reclaim expired sales") which is already reviewed
- Related but separate concern from the expiry bypass vulnerability

### False Report Risk
- Low - the changes are straightforward and well-commented
- The explicit comment at line 60-64 explains the intentional behavior

### Code Quality Notes
- No new tests visible in the diff - recommend adding tests for:
  - Overpayment refund behavior
  - ExactPrice plugin interaction with buy flow
- Consider edge case: what if refund `BankMsg::Send` fails?

### Status
**Security review complete.** This PR improves payment handling security. Not a fix for #65150 directly, but complements PR #118. Recommend merge after confirming test coverage.

---
*Automated security review by Crucible (2026-02-24)*

**Comment by crucible-burnt** [comment]
## 🔍 Crucible Security Review

### Summary
Two-part fix: adds ExactPrice plugin enforcement in `on_buy_plugin`, and adds defense-in-depth overpayment refund in `buy()`. Both changes improve economic security of the marketplace.

### Security Assessment
- **Risk Level**: Low (clean security improvement, one precision note)

**Refund logic is safe (`contracts/asset/src/execute/sale.rs:60–76`)**: `checked_sub` is used correctly after confirming `payment.amount.gt(&price.amount)`. CosmWasm message dispatch is post-execution, so no reentrancy concern with the `BankMsg::Send`.

**ExactPrice plugin wired into `on_buy_plugin` (`contracts/asset/src/plugin.rs:382–420`)**: The lookup uses a dummy `coin(0, "")` as the key — confirm this matches the actual storage key used when the plugin was originally set. If the plugin is keyed by denom+amount, a zero-coin lookup may silently miss it and treat it as absent.

**No access control gap**: the refund goes to `info.sender` (the buyer), which is correct.

### Immunefi Pattern Check
- This PR appears to address an overpayment / economic griefing issue. The fix pattern is correct.
- Potential false-report bait: the `coin(0, "")` plugin key lookup could be misread as a bug where the ExactPrice plugin is never found. A comment explaining that `get_plugin_name()` is key-derived (not amount-dependent) would head this off.

### False Report Risk
- The `coin(0, "")` lookup in `plugin.rs:385` will invite a report that ExactPrice is never enforced. Add a comment explaining that `get_plugin_name()` returns a name derived only from the plugin variant type, not the amount, so this is intentional.

### Code Quality Notes
- Old comment "check for underpayment but overpayment are absorbed if an exact price plugin is not set" was removed cleanly.

### Status
Conceptually sound. One clarification needed on the `coin(0, "")` key lookup before merge.

File: contracts/asset/src/execute/sale.rs [hunk 1]
```diff
@@ -58,6 +57,25 @@ where
 
     let mut response = Response::<TCustomResponseMsg>::default();
 
+    // defense-in-depth: cap payment to listing price and refund any excess
+    // to the buyer. The ExactPrice plugin rejects overpayment outright when
+    // configured; this ensures the seller never silently receives more than
+    // the listing price even when the plugin is absent.
+    if payment.amount.gt(&price.amount) {
```
[the remark this answers is not in the document]
Comment (2xburnt) [reply contracts/asset/src/execute/sale.rs:64]: Added focused overpayment coverage that asserts the refund precedes the net seller payment when deductions are present.

File: contracts/asset/src/plugin.rs [hunk 2]
```diff
@@ -379,6 +379,13 @@ where
     ) -> StdResult<bool> {
         // for buys we run the exact price, then allowed marketplaces and royalty plugins if set
         let config = AssetConfig::<TNftExtension>::default();
+        let exact_price_plugin = config.collection_plugins.may_load(
+            ctx.deps.storage,
+            Plugin::ExactPrice {
+                amount: cosmwasm_std::coin(0, ""),
+            }
+            .get_plugin_name(),
+        )?;
```
[the remark this answers is not in the document]
Comment (2xburnt) [reply contracts/asset/src/plugin.rs]: Replaced the dummy  construction with the static  storage key.

File: contracts/asset/src/execute/sale.rs [hunk 3]
```diff
@@ -58,6 +57,25 @@ where
...
+        let refund_amount = payment
+            .amount
+            .checked_sub(price.amount)
+            .map_err(|_| ContractError::InsufficientFunds {})?;
+        response = response.add_message(BankMsg::Send {
```
[the remark this answers is not in the document]
Comment (2xburnt) [reply contracts/asset/src/execute/sale.rs:69]: Changed royalty calculation to use the listing ask price, not the submitted overpayment, and added a regression test for an overpaid purchase.

File: contracts/asset/src/plugin.rs [hunk 4]
```diff
@@ -405,6 +412,9 @@ where
                 id: token_id.to_string(),
             })?;
         ctx.data.ask_price = Some(listing.price.clone());
+        if let Some(plugin) = exact_price_plugin {
+            plugin.run_asset_plugin(ctx)?;
+        }
```
[the remark this answers is not in the document]
Comment (2xburnt) [reply contracts/asset/src/plugin.rs]: Added a focused buy-hook test covering both a matching ExactPrice payment and rejected overpayment.

File: contracts/asset/src/plugin.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (2xburnt) [reply contracts/asset/src/plugin.rs]: Clarification: replaced the dummy Coin construction with the static ExactPrice storage key.

File: contracts/asset/src/plugin.rs [hunk 5]
```diff
@@ -405,6 +408,9 @@ where
...
```
[the remark this answers is not in the document]
Comment (2xburnt) [reply contracts/asset/src/plugin.rs]: Fixed in 7f7cedc: on_buy_plugin now runs the exact price check directly against the listing ask price instead of routing through run_asset_plugin, and the test configures a plugin amount different from the listing price to pin the behavior.
